### ISSUE TITLE
Missing $user->getUserFolder() in AppFramework

### ISSUE THREAD

**Issue by oparoz** [owncloud/core#17352]
We can do `$container->getUserFolder($userId)`, but that requires knowing about the user before creating the object which needs a folder.

If we had `$user->getUserFolder()`, we could inject the userManager, retrieve the user once we have its ID and then get the folder.

Makes sense @Raydiation?

**Comment by Raydiation** [comment]
I wouldn't solve it like this because a user should not be tied to the filesystem (Single Responsibility princible, see SOLID). It should work like this:

```
<?php 
class YourController {
    private $root;
    private $sharing;

    public function __construct($appName, IRequest $request, IRootFolder $root, ISharing $sharing) {
        parent::__construct($appName, $request);
        $this->root = $root;
        $this->sharing = $sharing;
    }

    public function doSomethingWithShareToken($token) {
        $user = $this->sharing->getUserFromToken($token);
        $folder = $this->root->getUserFolder($user);
    }
}
```

**Comment by oparoz** [comment]
The problem is this happens before the controller, in ISharing, in your example.
All the token is giving to the class is a userId and we need to derive the user and its folder from that in order to send the information to the controller or service when the need them.

So currently the whole container is injected in order to be able to call `getUserFolder`, but that's not appropriate.

It seems logical to be to be able to ask a user for his base folder. We currently can ask for its "home", but that's a string and we can't use it.

**Comment by Raydiation** [comment]
I dont get it, can you outline it? What should happen in what order?

**Comment by Raydiation** [comment]
ISharing (the class in question) just makes a database query where it fetches the user based on token, then returns an instantiated IUser object and you're done. It does not matter what happens before.

**Comment by oparoz** [comment]
Wait ISharing actually exists? I thought it was just a name used for the example.

**Comment by Raydiation** [comment]
Nope, just an example :)

**Comment by oparoz** [comment]
I can't find it...

So, here is the order in which things work.
The middleware gets a token and sends it to the environment class so that it can return the user and the folder to the controllers which need them.

The environment object is then injected in those controllers and services.

**Comment by Raydiation** [comment]
Why do you need a middleware and an environment?

**Comment by Raydiation** [comment]
If you are trying to log in a user based on the token this might be a potential security issue btw.

**Comment by oparoz** [comment]
In application
`$container->registerMiddleware('EnvCheckMiddleware');`

In middleware
$this->environment->setStandardEnv();

In environment
```
public function setTokenBasedEnv($linkItem) {
		// Resolves reshares down to the last real share
		$rootLinkItem = Share::resolveReShare($linkItem);
		$origShareOwner = $rootLinkItem['uid_owner'];
		$this->userFolder = $this->setupFilesystem($origShareOwner);
...
	}
```

In controller
`$this->environment->getNode($path)`

That will give me the content of the folder at that path.

That's the way the Files app is doing it, except it does it all in the controller, but using private APIs exclusively.

**Comment by oparoz** [comment]
The middleware is needed to do some basic checks before reaching the controllers

**Comment by Raydiation** [comment]
Yes, that logic is actually not needed at all. The following should just work (if the getUserFolder needs a user id like mentioned in the PR issue :))

```php
$rootLinkItem = Share::resolveReShare($linkItem);
$origShareOwner = $rootLinkItem['uid_owner'];
$this->userFolder = $this->root->getUserFolder($origShareOwner);
```

**Comment by Raydiation** [comment]
IMHO dont do a lot of stuff in your middleware, just check if things are authorized. Remember that middleware is run before every request.

**Comment by oparoz** [comment]
Ah, I think I understand the purpose of getRootFolder() now. I'm going to give it a go and report back. Thanks :)

**Comment by oparoz** [comment]
That was quick. That can't work because `IRootFolder` extends folder which doesn't have a `getUserFolder` method...

**Comment by Raydiation** [comment]
It's in a pull request

**Comment by Raydiation** [comment]
Basically has yet to be merged

**Comment by oparoz** [comment]
OK, so 8.2+ only. I'll make a note.

**Comment by Raydiation** [comment]
https://github.com/owncloud/core/pull/16965

**Comment by oparoz** [comment]
Thanks!

**Comment by oparoz** [comment]
Closed via #16965
